homeassistant/components/deconz.py

```diff
@@ -0,0 +1,134 @@
+"""Platform integrating Deconz support.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/deconz/
+"""
+
+import asyncio
+import logging
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_API_KEY, CONF_HOST, CONF_PASSWORD, CONF_PORT, CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP)
+from homeassistant.core import callback, EventOrigin
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.util import slugify
+from homeassistant.util.json import load_json, save_json
+
+REQUIREMENTS = ['pydeconz==10']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'deconz'
+
+DECONZ_DATA = 'deconz_data'
+CONFIG_FILE = 'deconz.conf'
+
+TYPE_AS_EVENT = 'type_as_event'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_PORT, default=80): cv.port,
+        vol.Optional(TYPE_AS_EVENT, default=['ZHASwitch']):
+            vol.All(cv.ensure_list, [cv.string]),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup services for Deconz."""
+    deconz_config = config[DOMAIN]
+    config_file = load_json(hass.config.path(CONFIG_FILE))
+
+    @callback
+    def _shutdown(call):  # pylint: disable=unused-argument
+        """Stop the connections to Deconz on shutdown."""
+        if DECONZ_DATA in hass.data:
+            _LOGGER.info("Stopping Deconz session.")
+            hass.data[DECONZ_DATA].close()
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+
+    @asyncio.coroutine
+    def generate_api_key(call):
+        """Generate API key needed to communicate with Deconz.
+
+        Store API key in deconz.conf.
+        """
+        from pydeconz.utils import get_api_key
+        deconz_config[CONF_USERNAME] = call.data.get(CONF_USERNAME, 'delight')
+        deconz_config[CONF_PASSWORD] = call.data.get(CONF_PASSWORD, 'delight')
+        api_key = yield from get_api_key(hass.loop, **deconz_config)
+        if not save_json(
+                hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key}):
+            _LOGGER.error("Failed to save API key to %s", CONFIG_FILE)
+        deconz_config[CONF_API_KEY] = api_key
+        yield from _setup_deconz(hass, config, deconz_config)
+
+    if CONF_API_KEY in deconz_config:
+        yield from _setup_deconz(hass, config, deconz_config)
+    elif CONF_API_KEY in config_file:
+        deconz_config[CONF_API_KEY] = config_file[CONF_API_KEY]
+        yield from _setup_deconz(hass, config, deconz_config)
+    else:
+        hass.services.async_register(
+            DOMAIN, 'generate_api_key', generate_api_key)
+        _LOGGER.warning("Deconz needs API key to set up session. See docs.")
```

Comment:
If the component can't do anything without an API key, this should probably be an error, not just a warning (i.e., return `False` here)

Reply:
An error is a bit extreme since you still intentionally have to enable the component. Which means that if you don't specify the API key you have the intent of generating one using the service which will only load when the API key is missing.

Reply:
Yes, I agree that a failure would be too hard if you have to call the `generate_api_key` service manually. But with the workflow from above, that should not be necessary, should it?

Reply:
Absolutely. See the other post for my reasoning.